trieChecker/cmd/main.go

```diff
@@ -0,0 +1,300 @@
+			}
+
+			directories = append(directories, c.Name())
+		}
+	}
+
+	numDirs := 0
+	for i := 0; i < maxDirs; i++ {
+		expectedDir := fmt.Sprintf("%d", i)
+		if !contains(directories, expectedDir) {
+			break
+		}
+
+		numDirs++
+	}
+
+	if numDirs == 0 {
+		return 0, fmt.Errorf("missing ordered directories in %s, like 0, 1 and so on", parentDir)
+	}
+	if numDirs != len(directories) {
+		return 0, fmt.Errorf("unordered directories in %s, like 0, 1 and so on", parentDir)
+	}
+
+	return numDirs - 1, nil
+}
+
+func contains(haystack []string, needle string) bool {
+	for _, h := range haystack {
+		if h == needle {
+			return true
+		}
+	}
+
+	return false
+}
+
+func checkTrie(flags config.ContextFlagsConfig, mainRootHash []byte, maxDBValue int) error {
+	marshaller := &marshal.GogoProtoMarshalizer{}
+	addressConverter, err := pubkeyConverter.NewBech32PubkeyConverter(addressLength, log)
+	if err != nil {
+		return err
+	}
+
+	tr, err := getTrie(marshaller, flags, maxDBValue)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		errNotCritical := tr.Close()
+		log.LogIfError(errNotCritical)
+	}()
+
+	ch, err := tr.GetAllLeavesOnChannel(mainRootHash)
+	if err != nil {
+		return err
+	}
+
+	numAccountsOnMainTrie := 0
+	numCodeNodes := 0
+	dataTriesRootHashes := make(map[string][]byte)
+	numDataTriesLeaves := 0
+	for kv := range ch {
+		numAccountsOnMainTrie++
+
+		userAccount := &state.UserAccountData{}
+		errUnmarshal := marshaller.Unmarshal(userAccount, kv.Value())
+		if errUnmarshal != nil {
+			// probably a code node
+			numCodeNodes++
+			continue
+		}
+
+		address := addressConverter.Encode(kv.Key())
+		dataTriesRootHashes[address] = userAccount.RootHash
+	}
+
+	log.Info("parsed main trie",
+		"num accounts", numAccountsOnMainTrie,
+		"num code nodes", numCodeNodes,
+		"num data tries", len(dataTriesRootHashes))
+
+	// TODO add error signaling in the trie implementation
+
+	for address, dataRootHash := range dataTriesRootHashes {
+		log.Debug("iterating data trie", "address", address, "data trie root hash", dataRootHash)
+
+		chDataTrie, errGetAllLeaves := tr.GetAllLeavesOnChannel(dataRootHash)
+		if errGetAllLeaves != nil {
+			return errGetAllLeaves
+		}
+
+		for range chDataTrie {
+			numDataTriesLeaves++
+		}
+	}
+
+	log.Info("parsed all tries",
+		"num accounts", numAccountsOnMainTrie,
+		"num code nodes", numCodeNodes,
+		"num data tries", len(dataTriesRootHashes),
+		"num data tries leaves", numDataTriesLeaves)
+
+	return nil
+}
+
+func getTrie(marshaller marshal.Marshalizer, flags config.ContextFlagsConfig, maxDBValue int) (common.Trie, error) {
+	hasher := blake2b.NewBlake2b()
+
+	cacheConfig := storageUnit.CacheConfig{
+		Type:        "SizeLRU",
+		Capacity:    500000,
+		SizeInBytes: 314572800, // 300MB
+	}
+	dbConfig := elrondConfig.DBConfig{
+		FilePath:          path.Join(flags.WorkingDir, flags.DbDir),
+		Type:              "LvlDBSerial",
+		BatchDelaySeconds: 2,
+		MaxBatchSize:      45000,
+		MaxOpenFiles:      10,
+	}
```

Comment:
Could you initialize these from a config file?

Reply:
this will make this tool harder to use as we will always need the config.toml file and its directory. Will extract them in a var file